### PR TITLE
do/#2725-parallel-tests

### PR DESCRIPTION
**Pull request by dimaopen**
Running tests in parallel: see the command line arguments in gradle spec task. 

There's still an issue: some services (i.e. the NetworkLoader) write data to the input dir (i.e. physsim-network.xml). This may cause Exceptions when several instances of BEAM loads data from the same input folder.

Resolved by #2725

---
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/2765)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -162,7 +162,7 @@ dependencies {
         exclude group: 'org.matsim', module: 'matsim'
     }
 
-    compile(group: 'com.github.wrashid.matsim', name: 'matsim', version: '0.10.1-beam-13') {
+    compile(group: 'org.matsim', name: 'matsim', version: '0.10.1-beam-14') {
```
[the remark this answers is not in the document]
Comment (JustinPihony) [reply build.gradle:165]: As well as this does not resolve currently - you'll need to push it to a public place

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java [hunk 2]
```diff
@@ -264,8 +262,6 @@ private void setupActorsAndRunPhysSim(IterationEndsEvent iterationEndsEvent) {
         completableFutures.add(CompletableFuture.runAsync(() -> physsimNetworkEuclideanVsLengthAttribute.notifyIterationEnds(iterationNumber)));
 
         writeIterationCsv(iterationNumber);
-        Road.setAllRoads(null);
```
Comment (JustinPihony) [inline src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java:267]: Why was this removed?

File: src/main/java/beam/physsim/jdeqsim/cacc/sim/JDEQSimulation.java [hunk 3]
```diff
@@ -52,7 +50,7 @@ protected void initializeVehicles() {
                 isCaccEnabled = value;
             }
             // the vehicle registers itself to the scheduler
-            new Vehicle(getScheduler(), person, activityDurationInterpretation, isCaccEnabled);
+            new Vehicle(getScheduler(), person, activityDurationInterpretation, isCaccEnabled, allRoads);
```
Comment (JustinPihony) [inline src/main/java/beam/physsim/jdeqsim/cacc/sim/JDEQSimulation.java:53]: Is this coming from JDEQSimulation? Is this part of the new matsim changes - can you link to that for review also? I know we try to not modify matsim too much to avoid merge issues later down the road

File: src/main/java/beam/physsim/jdeqsim/cacc/sim/Road.java [hunk 4]
```diff
@@ -191,23 +195,23 @@ public void procsessFilledRoad(org.matsim.core.mobsim.jdeqsim.Vehicle vehicle, d
...
 
         getDeadlockPreventionMessages().add(vehicle.scheduleDeadlockPreventionMessage(timeToLeaveRoad, this));
 
         assert (getInterestedInEnteringRoad().size()== getDeadlockPreventionMessages().size()) :getInterestedInEnteringRoad().size() + " - " + getDeadlockPreventionMessages().size();
     }
 
-    public static Road getRoad(Id<Link> linkId) {
-        return (Road) getAllRoads().get(linkId);
+    public Road getRoad(Id<Link> linkId) {
+        return (Road) allRoads.get(linkId);
```
Comment (JustinPihony) [inline src/main/java/beam/physsim/jdeqsim/cacc/sim/Road.java:215]: Again - maybe this is managed more in matsim now, but I would think that we would want to sync with the underlying mechanism and keep using the `getAllRoads`

File: src/main/scala/beam/analysis/cartraveltime/CarTripStatsFromPathTraversalEventHandler.scala [hunk 5]
```diff
@@ -331,33 +332,6 @@ class CarTripStatsFromPathTraversalEventHandler(
     )
   }
 
-  /**
-    * Plots graph for average travel times per hour at iteration level
-    *
-    * @param dataset category dataset for graph genration
-    * @param iterationNumber iteration number
-    */
-  private def createIterationGraphForAverageCarTravelTime(dataset: CategoryDataset, iterationNumber: Int): Unit = {
```
Comment (JustinPihony) [inline src/main/scala/beam/analysis/cartraveltime/CarTripStatsFromPathTraversalEventHandler.scala:340]: I'm using the github UI - was this unused and that was why it was removed?

File: src/main/scala/beam/router/skim/Skims.scala [hunk 6]
```diff
@@ -2,43 +2,46 @@ package beam.router.skim
...
+  skims.put(SkimType.DT_SKIMMER, addEvent(driveTimeSkimmer))
 
+  private def addEvent(skimmer: AbstractSkimmer): AbstractSkimmer = {
+    matsimServices.addControlerListener(skimmer)
+    matsimServices.getEvents.addHandler(skimmer)
+    skimmer
   }
 
-  def clear(): Unit = {
-    skims.clear()
```
Comment (JustinPihony) [inline src/main/scala/beam/router/skim/Skims.scala:24]: Was there a need for this clear? It seems like it would have been a cleanup needed elsewhere?

File: src/main/scala/beam/sim/BeamHelper.scala [hunk 7]
```diff
@@ -454,7 +459,7 @@ trait BeamHelper extends LazyLogging {
     }), scala.concurrent.duration.Duration.Inf)
   }
 
-  def runBeamWithConfig(config: TypesafeConfig): (MatsimConfig, String) = {
+  def runBeamWithConfig(config: TypesafeConfig): (MatsimConfig, String, BeamServices) = {
```
Comment (JustinPihony) [inline src/main/scala/beam/sim/BeamHelper.scala:462]: I probably missed it, but this was added and only seems to be ignored? What was the purpose of adding it?

File: src/main/scala/beam/router/skim/Skims.scala (hunk 6 above)
Comment (JustinPihony) [reply src/main/scala/beam/router/skim/Skims.scala:24]: Looking at its usage it seems only because it was a static object and this was a hack?

File: src/test/java/beam/analysis/plots/DeadHeadingGraphTest.java [hunk 8]
```diff
@@ -9,7 +9,7 @@
 import static org.junit.Assert.assertEquals;
 
 public class DeadHeadingGraphTest {
-    private final DeadHeadingAnalysis deadHeadingStats = new DeadHeadingAnalysis(NoOpSimulationMetricCollector$.MODULE$, true);
+    private final DeadHeadingAnalysis deadHeadingStats = new DeadHeadingAnalysis(NoOpSimulationMetricCollector$.MODULE$, true, null);
```
Comment (JustinPihony) [inline src/test/java/beam/analysis/plots/DeadHeadingGraphTest.java:12]: nulls scare me a bit. Should we make this an Optional at least - or fill it with a mocked object?

File: src/test/scala/beam/agentsim/agents/household/FastHouseholdCAVSchedulingSpec.scala [hunk 9]
```diff
@@ -64,7 +64,6 @@ class FastHouseholdCAVSchedulingSpec
 
   describe("A Household CAV Scheduler") {
     it("generates two schedules") {
-      Skims.setup(services)
```
Comment (JustinPihony) [inline src/test/scala/beam/agentsim/agents/household/FastHouseholdCAVSchedulingSpec.scala:67]: Where is the injection set up in the tests?

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java (hunk 2 above)
Comment (dimaopen) [reply src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java:267]: This line clears a static field for I suppose freeing the memory. Because that static field is removed (that field is moved to Vehicle) we don't more need to explicitly clear it. It is disposed by GC when it frees memory automatically.

File: build.gradle [hunk 10]
```diff
@@ -340,7 +344,7 @@ build.dependsOn spec
 
 task taggedTest(dependsOn: ['testClasses'], type: JavaExec) {
     main = 'org.scalatest.tools.Runner'
-    args = ['-R', 'build/classes/scala/test', '-o', '-n'] << (project.findProperty('tags') ?: 'org.scalatest.Ignore')
+    args = ['-R', 'build/classes/scala/test', '-o', '-P4', '-n'] << (project.findProperty('tags') ?: 'org.scalatest.Ignore')
```
[the remark this answers is not in the document]
Comment (dimaopen) [reply build.gradle:347]: I removed this flag because using tagged tests implies a special scenario where parallel execution can be unwanted.

File: src/main/java/beam/physsim/jdeqsim/cacc/sim/JDEQSimulation.java (hunk 3 above)
Comment (dimaopen) [reply src/main/java/beam/physsim/jdeqsim/cacc/sim/JDEQSimulation.java]: That's correct. Though without this change (removing static fields) we cannot run tests in the same jvm. I had to change matsim. Though these changes aren't so big.

File: src/main/java/beam/physsim/jdeqsim/cacc/sim/Road.java (hunk 4 above)
Comment (dimaopen) [reply src/main/java/beam/physsim/jdeqsim/cacc/sim/Road.java:215]: Another solution would be running beam and tests using different classloaders. I have not researched this possibility yet.

File: src/main/scala/beam/analysis/cartraveltime/CarTripStatsFromPathTraversalEventHandler.scala (hunk 5 above)
Comment (dimaopen) [reply src/main/scala/beam/analysis/cartraveltime/CarTripStatsFromPathTraversalEventHandler.scala:340]: Yes, that's correct. Though if this method is going to be needed soon I can restore it and handle the static field somehow.

File: src/main/scala/beam/router/skim/Skims.scala (hunk 6 above)
Comment (dimaopen) [reply src/main/scala/beam/router/skim/Skims.scala:24]: Yes, they clear this because it's a static object and they need to explicitly clear it to free the memory.

File: src/main/scala/beam/sim/BeamHelper.scala (hunk 7 above)
Comment (dimaopen) [reply src/main/scala/beam/sim/BeamHelper.scala:462]: I added BeamServices to the return value in order to get them in ExperimentRunner#runExperiment. These services are needed for getting the output directory for evaluation.
https://github.com/LBNL-UCB-STI/beam/blob/051113d0cb20df7ceb1ce6c3ef2bfe30c64e2fb9/src/main/scala/beam/calibration/ExperimentRunner.scala#L53-L63

File: src/test/java/beam/analysis/plots/DeadHeadingGraphTest.java (hunk 8 above)
Comment (dimaopen) [reply src/test/java/beam/analysis/plots/DeadHeadingGraphTest.java:12]: I created a mock for this test. At least it provides a meaningful exeception instead of an NPE if it is called.

File: src/test/scala/beam/agentsim/agents/household/FastHouseholdCAVSchedulingSpec.scala (hunk 9 above)
Comment (dimaopen) [reply src/test/scala/beam/agentsim/agents/household/FastHouseholdCAVSchedulingSpec.scala:67]: Now it is done in a regular Guice way during module definition in BeamHelper. https://github.com/LBNL-UCB-STI/beam/blob/4a4014f7e2df10372c45e03dfa7605ba7ad128aa/src/main/scala/beam/sim/BeamHelper.scala#L218

File: build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (dimaopen) [reply build.gradle:165]: That' correct. I created a PR for matsim. https://github.com/wrashid/matsim/pull/9. For some reason, the branch which I branch off has this artifact group in the classifier.
I'll change it and push the artifact to a public repo during working on that PR.

File: src/main/java/beam/physsim/jdeqsim/cacc/sim/Road.java (hunk 4 above)
Comment (JustinPihony) [reply src/main/java/beam/physsim/jdeqsim/cacc/sim/Road.java:215]: After looking at the other PR I think this seems to be more semantic than anything else....although the underlying allRoads issue is up for grabs there.

**Comment by JustinPihony** [comment]
test!

**Comment by JustinPihony** [comment]
test!

**Comment by JustinPihony** [comment]
test!

**Comment by JustinPihony** [comment]
test!

**Comment by dimaopen** [comment]
test!

**Comment by dimaopen** [comment]
Test!

**Comment by dimaopen** [comment]
Test!

**Comment by dimaopen** [comment]
Test!

**Comment by JustinPihony** [comment]
test!

**Comment by JustinPihony** [comment]
Test!
